Make hash-table support numbers as keys

nagy opened the pull request:
This allows numbers to be used as hash-table keys, since they are crudely converted to strings.

Closes #424

vlad-km commented:
@nagy, thanks.
Sorry, very busy lately.
I'll take a look at the code this weekend.

nagy commented:
> that (gethash 1 ht) should be different from (gethash "1" ht)

You are right, I will think of a different approach.

hemml commented:
May be we can store keys as strings like "TYPE:VALUE", for example "1" will be converted to "string:1" and 1 will be "number:1", etc.

nagy commented:
To my surprise, that distinction between numbers and strings is already happening correctly. I have split the commit into two and added some more test and wrote some reasoning into the commit message.

Please check if it is fine now.

davazp commented:
Awesome!  Thanks!
